src/main/java/graphql/normalized/ValueToVariableValueCompiler.java

```diff
@@ -0,0 +1,128 @@
+            Value<?> value = (Value<?>) maybeValue;
+            variableValue = toVariableValue(value);
+        } else if (maybeValue instanceof List) {
+            variableValue = normalisedValueToVariableValues((List<Object>) maybeValue);
+        } else if (maybeValue instanceof Map) {
+            variableValue = normalisedValueToVariableValues((Map<String, Object>) maybeValue);
+        } else {
+            throw new AssertException("Should never happen. Did not expect type: " + maybeValue.getClass());
+        }
+        return variableValue;
+    }
+
+    private static List<Object> normalisedValueToVariableValues(List<Object> arrayValues) {
+        return map(arrayValues, ValueToVariableValueCompiler::normalisedValueToVariableValue);
+    }
+
+    @NotNull
+    private static Map<String, Object> normalisedValueToVariableValues(Map<String, Object> objectMap) {
+        Map<String, Object> output = new LinkedHashMap<>();
+        objectMap.forEach((k, v) -> {
+            Object value = normalisedValueToVariableValue(v);
+            output.put(k, value);
+        });
+        return output;
+    }
+
+    private static Map<String, Object> toVariableValue(ObjectValue objectValue) {
+        Map<String, Object> map = new LinkedHashMap<>();
+        List<ObjectField> objectFields = objectValue.getObjectFields();
+        for (ObjectField objectField : objectFields) {
+            String objectFieldName = objectField.getName();
+            Value<?> objectFieldValue = objectField.getValue();
+            map.put(objectFieldName, toVariableValue(objectFieldValue));
+        }
+        return map;
+    }
+
+    @NotNull
+    private static List<Object> toVariableValues(List<Value> arrayValues) {
+        // some values can be null (NullValue) and hence we can use Immutable Lists
+        return arrayValues.stream()
+                .map(ValueToVariableValueCompiler::toVariableValue)
+                .collect(toList());
+    }
+
+    @Nullable
+    private static Object toVariableValue(Value<?> value) {
+        if (value instanceof ObjectValue) {
+            return toVariableValue((ObjectValue) value);
+        } else if (value instanceof ArrayValue) {
+            return toVariableValues(((ArrayValue) value).getValues());
+        } else if (value instanceof StringValue) {
+            return ((StringValue) value).getValue();
+        } else if (value instanceof FloatValue) {
+            return ((FloatValue) value).getValue();
+        } else if (value instanceof IntValue) {
+            return ((IntValue) value).getValue();
+        } else if (value instanceof BooleanValue) {
+            return ((BooleanValue) value).isValue();
+        } else if (value instanceof NullValue) {
+            return null;
+        }
+        throw new AssertException("Should never happen. Cannot handle node of type: " + value.getClass());
+    }
+
+    private static String getVarName(int variableOrdinal) {
+        return "v" + variableOrdinal;
+    }
```

Comment:
I shortened the original work that Artyom put in as `var_0` - I figure we should be as short as possible